File: dms/api/service_estimates.py
"""Service Estimate API for the DMS frontend."""

import frappe
from frappe import _
from frappe.utils import flt

from dms.api.utils import LIST_ORDER_LATEST_CREATED, add_branch_filter, get_dms_companies
from dms.dealer_management_system.doctype.dms_service_estimate.estimate_utils import (
	sync_job_card_from_accepted_estimate,
)
from dms.dealer_management_system.utils.document_links import (
	enrich_estimate_row,
	linked_job_card_for_estimate,
)


def _customer_display_name(customer):
	if not customer:
		return None
	return frappe.db.get_value("Customer", customer, "customer_name")


@frappe.whitelist()
def get_service_estimates(limit=50, offset=0, status=None, customer=None, search=None):
	filters = {}
	if status:
		filters["status"] = status
	if customer:
		filters["customer"] = customer

	companies = get_dms_companies()
	if companies:
		filters["company"] = ["in", companies]

	or_filters = {}
	if search:
		or_filters = {
			"name": ["like", f"%{search}%"],
			"customer": ["like", f"%{search}%"],
			"license_plate": ["like", f"%{search}%"],
			"vehicle_vin": ["like", f"%{search}%"],
		}

	filters = add_branch_filter(filters, doctype="DMS Service Estimate")

	total = len(
		frappe.get_all(
			"DMS Service Estimate",
			filters=filters,
			or_filters=or_filters or None,
			limit_page_length=0,
			pluck="name",
		)
	)

	rows = frappe.get_all(
		"DMS Service Estimate",
		filters=filters,
		or_filters=or_filters or None,
		fields=[
			"name",
			"status",
			"customer",
			"customer_name",
			"vehicle_vin",
			"license_plate",
			"inspection",
			"appointment",
			"diagnostic_invoice",
			"diagnostic_fee",
			"total_before_vat",
			"grand_total",
			"customer_decision",
			"company",
			"posting_date",
			"creation",
			"modified",
		],
		limit=int(limit),
		limit_start=int(offset),
		order_by=LIST_ORDER_LATEST_CREATED,
	)

	for row in rows:
		enrich_estimate_row(row)

	return {"data": rows, "total": total}


@frappe.whitelist()
def get_service_estimate(name):
	if not name:
		frappe.throw(_("Service Estimate name is required"))

	doc = frappe.get_doc("DMS Service Estimate", name)
	doc.check_permission("read")

	result = doc.as_dict()
	result["customer_name"] = result.get("customer_name") or _customer_display_name(doc.customer)
	if doc.vehicle_vin:
		from dms.api.service_packages import resolve_vehicle_model_from_vin

		vm, vm_label = resolve_vehicle_model_from_vin(doc.vehicle_vin)
		result["vehicle_model"] = vm
		result["vehicle_model_label"] = vm_label
	enrich_estimate_row(result)
	if doc.appointment:
		result["assigned_bay"] = frappe.db.get_value(
			"Service Appointment", doc.appointment, "assigned_bay"
		)
	return result


@frappe.whitelist()
def get_customer_terms_and_conditions():
	"""English + Arabic customer terms for estimate approval."""
	from dms.api.common import get_customer_terms_and_conditions as _get

	return _get()


@frappe.whitelist()
def update_service_estimate(name, data):
	if isinstance(data, str):
		import json

		data = json.loads(data)

	doc = frappe.get_doc("DMS Service Estimate", name)
	doc.check_permission("write")

	if doc.status in ("Rejected", "Cancelled"):
		frappe.throw(_("This estimate can no longer be edited."))

	was_accepted = doc.status == "Accepted"

	allowed_child = {"labour", "parts"}
	scalar_fields = {
		"diagnosis_findings",
		"recommended_repairs",
		"service_advisor_notes",
		"internal_notes",
		"service_package",
		"vat_rate",
		"status",
		"warranty_application_type",
		"labour_discount_type",
		"labour_discount_value",
		"parts_discount_type",
		"parts_discount_value",
	}

	for field in scalar_fields:
		if field in data:
			doc.set(field, data[field])

	from dms.dealer_management_system.doctype.dms_job_card.job_card_discount import (
		apply_discount_fields_from_payload,
	)

	if any(k in data for k in ("labour_discount", "parts_discount")):
		apply_discount_fields_from_payload(doc, data)

	for table in allowed_child:
		if table in data and isinstance(data[table], list):
			doc.set(table, [])
			for row in data[table]:
				doc.append(table, row)

	doc.save()
	synced_job_card = None
	if was_accepted:
		synced_job_card = sync_job_card_from_accepted_estimate(doc)
	frappe.db.commit()

	result = doc.as_dict()
	result["customer_name"] = result.get("customer_name") or _customer_display_name(doc.customer)
	if synced_job_card:
		result["synced_job_card"] = synced_job_card
	enrich_estimate_row(result)
	return result


@frappe.whitelist()
def delete_service_estimate(name):
	if not name:
		frappe.throw(_("Service Estimate name is required"))

	doc = frappe.get_doc("DMS Service Estimate", name)
	doc.check_permission("delete")

	linked_jc = linked_job_card_for_estimate(doc.name)
	if linked_jc and frappe.db.exists("DMS Job Card", linked_jc):
		frappe.throw(
			_("Cannot delete — linked job card {0} exists.").format(frappe.bold(linked_jc))
		)

	if doc.diagnostic_invoice and frappe.db.exists("Sales Invoice", doc.diagnostic_invoice):
		frappe.throw(
			_("Cannot delete — diagnostic invoice {0} exists.").format(
				frappe.bold(doc.diagnostic_invoice)
			)
		)

	frappe.delete_doc("DMS Service Estimate", name, force=1)
	frappe.db.commit()
	return {"deleted": name}


@frappe.whitelist()
def get_dms_estimate_settings():
	return {
		"default_diagnostic_fee": flt(
			frappe.db.get_single_value("DMS Settings", "default_diagnostic_fee") or 3000
		),
		"default_vat_rate": flt(frappe.db.get_single_value("DMS Settings", "default_vat_rate") or 15),
		"diagnostic_fee_item": frappe.db.get_single_value("DMS Settings", "diagnostic_fee_item"),
	}

File: frontend/services/serviceEstimates.ts
/**
 * DMS Service Estimate — diagnosis → estimation → customer approval workflow
 */
import { apiRequest } from './apiClient';
import { fetchCustomerTermsAndConditions } from './common';
import type { DMSServiceEstimate, PaginatedResponse } from '@/types/dms';

const API = 'dms.api.service_estimates';
const DT = 'dms.dealer_management_system.doctype.dms_service_estimate.dms_service_estimate';

export async function listServiceEstimates(options?: {
  status?: string;
  customer?: string;
  search?: string;
  limit?: number;
  offset?: number;
}): Promise<PaginatedResponse<DMSServiceEstimate>> {
  return apiRequest<PaginatedResponse<DMSServiceEstimate>>(
    `/api/method/${API}.get_service_estimates`,
    {
      method: 'POST',
      body: JSON.stringify({
        status: options?.status || null,
        customer: options?.customer || null,
        search: options?.search || null,
        limit: options?.limit || 50,
        offset: options?.offset || 0,
      }),
    }
  );
}

export async function getServiceEstimate(name: string): Promise<DMSServiceEstimate> {
  return apiRequest<DMSServiceEstimate>(`/api/method/${API}.get_service_estimate`, {
    method: 'POST',
    body: JSON.stringify({ name }),
  });
}

export async function updateServiceEstimate(
  name: string,
  data: Partial<DMSServiceEstimate>
): Promise<DMSServiceEstimate> {
  return apiRequest<DMSServiceEstimate>(`/api/method/${API}.update_service_estimate`, {
    method: 'POST',
    body: JSON.stringify({ name, data }),
  });
}

export async function deleteServiceEstimate(name: string): Promise<{ deleted: string }> {
  return apiRequest<{ deleted: string }>(`/api/method/${API}.delete_service_estimate`, {
    method: 'POST',
    body: JSON.stringify({ name }),
  });
}

export async function getEstimateSettings(): Promise<{
  default_diagnostic_fee: number;
  default_vat_rate: number;
  diagnostic_fee_item?: string;
}> {
  return apiRequest(`/api/method/${API}.get_dms_estimate_settings`, {
    method: 'POST',
    body: JSON.stringify({}),
  });
}

export async function makeFromInspection(inspection: string): Promise<string> {
  return apiRequest<string>(`/api/method/${DT}.make_service_estimate_from_inspection`, {
    method: 'POST',
    body: JSON.stringify({ source_name: inspection }),
  });
}

export async function completeDiagnosis(
  estimateName: string,
  payload: { diagnosis_findings?: string; recommended_repairs?: string }
): Promise<{ name: string; status: string }> {
  return apiRequest(`/api/method/${DT}.complete_diagnosis`, {
    method: 'POST',
    body: JSON.stringify({
      estimate_name: estimateName,
      diagnosis_findings: payload.diagnosis_findings ?? null,
      recommended_repairs: payload.recommended_repairs ?? null,
    }),
  });
}

export async function startEstimation(estimateName: string): Promise<{ name: string; status: string }> {
  return apiRequest(`/api/method/${DT}.start_estimation`, {
    method: 'POST',
    body: JSON.stringify({ estimate_name: estimateName }),
  });
}

export async function submitForCustomerApproval(estimateName: string): Promise<{
  name: string;
  status: string;
  total_before_vat: number;
  grand_total: number;
}> {
  return apiRequest(`/api/method/${DT}.submit_for_customer_approval`, {
    method: 'POST',
    body: JSON.stringify({ estimate_name: estimateName }),
  });
}

export async function getCustomerTermsAndConditions() {
  return fetchCustomerTermsAndConditions();
}

export async function acceptEstimate(
  estimateName: string,
  payload: {
    customer_signature: string;
    lead_technician?: string;
    assigned_bay?: string;
    schedule_start_time?: string;
    schedule_end_time?: string;
    start_repair?: boolean;
    terms_accepted?: boolean;
    job_card_type?: string;
  }
): Promise<{ name: string; status: string; job_card: string }> {
  return apiRequest(`/api/method/${DT}.accept_estimate`, {
    method: 'POST',
    body: JSON.stringify({
      estimate_name: estimateName,
      customer_signature: payload.customer_signature,
      lead_technician: payload.lead_technician || null,
      assigned_bay: payload.assigned_bay || null,
      schedule_start_time: payload.schedule_start_time || null,
      schedule_end_time: payload.schedule_end_time || null,
      start_repair: payload.start_repair ? 1 : 0,
      terms_accepted: payload.terms_accepted ? 1 : 0,
      job_card_type: payload.job_card_type || null,
    }),
  });
}

export async function rejectEstimate(
  estimateName: string,
  rejectionSignature: string,
  termsAccepted = false,
  options?: {
    rejection_reason?: string;
    lost_sale_follow_up_date?: string;
  }
): Promise<{ name: string; status: string; diagnostic_invoice: string }> {
  return apiRequest(`/api/method/${DT}.reject_estimate`, {
    method: 'POST',
    body: JSON.stringify({
      estimate_name: estimateName,
      rejection_signature: rejectionSignature,
      terms_accepted: termsAccepted ? 1 : 0,
      rejection_reason: options?.rejection_reason || null,
      lost_sale_follow_up_date: options?.lost_sale_follow_up_date || null,
    }),
  });
}
